app/buttle/action.py

```diff
@@ -0,0 +1,66 @@
+from __future__ import annotations
+from app.buttle.models import Knight, Weapon, Effect, Potion, Armour
+
+
+def make_knights(knights_config: dict) -> list[Knight]:
+    knights_instances = []
+
+    for knight_name in knights_config:
+        knight_dict = knights_config[knight_name]
+
+        armours_list = []
+        if knight_dict.get("armour"):
+            armours_list = make_armour(
+                knight_dict.get("armour"))
+
+        weapon = Weapon(name=knight_dict["weapon"]["name"],
+                        power=knight_dict["weapon"]["power"])
+
+        potion = None
+        if knight_dict.get("potion"):
+            potion = make_potion(knight_dict.get("potion"))
+
+        new_knight = Knight(name=knight_dict["name"],
+                            power=knight_dict["power"],
+                            hp=knight_dict["hp"],
+                            armour=armours_list,
+                            weapon=weapon,
+                            potion=potion)
+        new_knight.protection = 0
+
+        if potion:
+            new_knight.potion_bonus(potion)
+
+        new_knight.calculate_protection()
+        new_knight.calculate_power()
+        knights_instances.append(new_knight)
```

Comment:
try to remove this logic to the Knight instance methods, so u just need to create an instances

Reply:
Moved the logics to classes .